cloud/scope/machinepool.go

```diff
@@ -138,7 +138,9 @@ func (m *MachinePoolScope) ProviderID() string {
 
 // SetProviderID sets the AzureMachinePool providerID in spec.
 func (m *MachinePoolScope) SetProviderID(v string) {
-	m.AzureMachinePool.Spec.ProviderID = v
+	if m.AzureMachinePool.Spec.ProviderID != "" {
```

Comment:
I just realized this by looking at another issue, I think we need to keep the overwrite. I know it seems wrong since providerID for a VM should never change but since this field is part of spec it is technically modifiable by the user and so I think the idea is we should always overwrite it with the source of truth. We can't set up the webhook to reject updates to providerID either because otherwise the controller will fail when trying to set it for the first time.

It should probably be labeled as "READ-ONLY" in the CRD, too.

Reply:
Makes sense. I had received mixed guidance on this when submitting a PR to CAPI. I like your reasoning. Also, with patch helper, it should only be updated if it has changed.